pyro/infer/autoguide/guides.py

```diff
@@ -656,44 +657,75 @@ def _loc_scale(self, *args, **kwargs):
         return self.loc, scale
 
 
-class AutoIAFNormal(AutoContinuous):
+class AutoNormalizingFlow(AutoContinuous):
     """
     This implementation of :class:`AutoContinuous` uses a Diagonal Normal
-    distribution transformed via a :class:`~pyro.distributions.transforms.AffineAutoregressive`
-    to construct a guide over the entire latent space. The guide does not depend on the model's
-    ``*args, **kwargs``.
+    distribution transformed via one of the normalizing flows in
+    :mod:`~pyro.distributions.transforms`to construct a guide over the entire
+    latent space. The guide does not depend on the model's ``*args, **kwargs``.
 
     Usage::
 
-        guide = AutoIAFNormal(model, hidden_dim=latent_dim)
+        guide = AutoNormalizingFlow(model, 2, affine_autoregressive)
         svi = SVI(model, guide, ...)
 
     :param callable model: a generative model
-    :param int hidden_dim: number of hidden dimensions in the IAF
+    :param num_flows: number of normalizing flows
+    :param init_transform_fn: a callable which when provided with the latent
+        dimension and additional keyword arguments (`init_transform_kwargs`)
+        returns an instance of
+        :class:`~pyro.distributions.torch_tranform.TransformModule`.
     :param callable init_loc_fn: A per-site initialization function.
         See :ref:`autoguide-initialization` section for available functions.
+    :param init_tranform_kwargs: other keyword arguments taken by
+        `init_transform_fn`.
     """
 
-    def __init__(self, model, hidden_dim=None, init_loc_fn=init_to_median):
-        self.hidden_dim = hidden_dim
-        self.arn = None
+    def __init__(self, model, num_flows, init_transform_fn, init_loc_fn=init_to_median,
+                 **init_transform_kwargs):
+        self.num_flows = num_flows
+        self.transform = None
+        self._init_transform_fn = init_transform_fn
```

Comment:
i think the general idea of using these pre-made factories is a good one :+1:

Reply:
Normalizing flows usually contain a number of different transforms of different types. You can compose any of the transforms in the library into a normalizing flow...

Does it make sense then to pass a single transform function that can be repeated? If you wanted a more complex flow, you'd have to write a helper function, something like:

```
def make_transforms(latent_dim=1):
  transforms = [T.affine_autoregressive(...), T.batch_norm(...), T.affine_autoregressive(...)]
  return T.ComposeTransforms(transforms)
```

Reply:
Also, not all transforms have helper functions (like the ones from PyTorch, and a few that I didn't add myself to Pyro...)

Reply:
@stefanwebb - I added `num_flows` since that seemed like a common use case. Many experiments that I have seen just use a single transform repeated n times, so this is to accommodate for that use case without having to write custom helper functions. For more complex combinations, I foresee users using a `make_transforms` function like you suggest.

Reply:
@neerajprad re `num_flows`: I agree with @stefanwebb and would aim to "program with functions" rather than "with *args,**kwargs" if possible. You might replace the `num_flows` arg with a new higher-order helper function like
```py
def iterated(repeats, base_fn, *args, **kwargs):
    assert isinstance(repeats, int) and repeats >= 0
    return ComposeTransformModule([base_fn(*args, **kwargs)
                                   for _ in range(repeats)])
```
That way I think you could replace `num_flows` with something like
```diff
- AutoNormalizingFlow(init_transform_fn, num_flows)
+ AutoNormalizingFlow(partial(iterated, num_flows, init_transform_fn))
```

Reply:
Done.